### DISCUSSION TITLE
Game, spells that cannot be cast, make a spell that cannot be cast for whatever reason "greyed out" (the spell text) in the Magic Book

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#4845] [Ideas]
To avoid this "That spell will affect no one!" window, I suggest making a spell that cannot be cast for whatever reason "greyed out" (the spell text) in the Magic Book in the same way that a spell is greyed out when the hero does not have enough spell points to cast the spell, and that therefore it is not possible to cast the spell.
The spell information window should also tell the player why it is not possible to cast the spell.

With this change, this one would no longer be necessary: https://github.com/ihhub/fheroes2/issues/1704.

In fheroes2:

The spell cannot be cast.
![Spells that cannot be cast 02](https://user-images.githubusercontent.com/43583994/93310903-a41b5100-f805-11ea-915a-f1a6a67e68f0.png)

The spell's "greyed out" text in the Magic Book.
![Greyed spell 001](https://user-images.githubusercontent.com/43583994/145685229-478acc44-930c-4332-838c-35f5562e97cc.png)

This affects all combat spells, depending on the case.

There are special cases for some combat spells, and also depending on the creature:

- There is no advantage to cast the "Bless" spell, or the "Curse" spell, ~or the "Slow" spell~ if the only troop is a troop of Peasants, so it should not be possible to cast either of these spells in this case.
- There is no advantage to cast the "Shield" spell if there are no shooters in the opposing troops, so it should not be possible to cast this spell in this case.
- There is no advantage to cast the "Disrupting Ray" spell if all opposing troops are already at the minimum possible defense, so it should not be possible to cast this spell in this case.
- There is no advantage to cast the "Earthquake" spell if all parts of the castle are already destroyed, so it should not be possible to cast this spell in this case.
- There is no advantage to cast the "Dragon Slayer" spell if there are no Dragons in the opposing troops, so it should not be possible to cast this spell in this case.

**Changes for all adventure spells have already been done.**

For adventure map spells: **[Done]**

- The case of the "Town Gate" spell is already covered here: https://github.com/ihhub/fheroes2/issues/4771.
- It should not be possible to cast the "Haunt" spell or one of the "Set Elemental Guardian" spells if the hero is not positioned on a mine, or if the spell is already active and re-casting the spell would not be beneficial (less creatures generated).
- It should not be possible to cast the "Summon Boat" spell if the hero is not positioned near the ocean, or if there is no boat available, or if there is no more free adjacent square to host the boat.
Related discussion: https://github.com/ihhub/fheroes2/discussions/2593.
- It should not be possible to cast the "Visions" spell if there are no wandering troops within the hero's vision radius.
- It should not be possible to cast the "Identify Hero" spell if there are no opposing heroes on the adventure map, or if the spell is already active.
- It should not be possible to cast the "Town Portal" spell if the player does not have a town/castle, or if all other towns/castles are already occupied by other visiting heroes.

**Comment by Kazhuu** [comment]
From adventure spells Summon Boat, Town Gate and Town Portal are already having grayed out behavior. This is done in this PR https://github.com/ihhub/fheroes2/pull/6729.

**Comment by Kazhuu** [comment]
> There is no advantage to cast the "Bless" spell, or the "Curse" spell, or the "Slow" spell if the only troop is a troop of Peasants, so it should not be possible to cast either of these spells in this case.

Should this also applied to mass version of these spells? Then I was thinking other spells as well like Dispel. Should it be not possible to cast this as well if there are not creatures affected by any spells? Then how about spell Holy Word if there no undead in the battle?

**Reply by LeHerosInconnu** [reply]
Hello @Kazhuu,

> > There is no advantage to cast the "Bless" spell, or the "Curse" spell, or the "Slow" spell if the only troop is a troop of Peasants, so it should not be possible to cast either of these spells in this case.
> 
> Should this also applied to mass version of these spells?

Yes, the mass spells of these spells are also concerned.

> Then I was thinking other spells as well like Dispel. Should it be not possible to cast this as well if there are not creatures affected by any spells? Then how about spell Holy Word if there no undead in the battle?

It is the special cases that are listed in the original post.

"This affects all combat spells, depending on the case.

There are special cases for some combat spells, and also depending on the creature:"

For other spells, as soon as the message "That spell will affect no one!" should be displayed, the spell should be grayed out in the Magic Book.

For example, if the hero tries to cast the Resurrection spell when there are no creatures to resurrect in his army, the Magic Book is closed and the message "That spell will affect no one!" (with an "OKAY" button) is displayed on the battlefield screen.
Instead of the message, the spell should be directly grayed out in the Magic Book when it cannot be used.
And if the hero tries to cast the grayed spell anyway, the message "That spell will affect no one!" (with an "OKAY" button) is displayed directly in the Magic Book.

**Comment by Kazhuu** [comment]
Hello @LeHerosInconnu! I think I fixed all the adventure spells. Should this discussion be updated to reflect that they are done. Mostly because if someone else is reading this and want to contribute as well. Now the information on this discussion is bit misleading. 

Also I'm now working on combat spells. Maybe we want to change some of the messages to show why you cannot cast the spell. Not just displaying "That spell will affect no one!" in every case, right?

**Reply by LeHerosInconnu** [reply]
Hello @Kazhuu,

> Hello @LeHerosInconnu! I think I fixed all the adventure spells. Should this discussion be updated to reflect that they are done. Mostly because if someone else is reading this and want to contribute as well. Now the information on this discussion is bit misleading.

I will indicate in the original message that this has been done.

> Also I'm now working on combat spells. Maybe we want to change some of the messages to show why you cannot cast the spell. Not just displaying "That spell will affect no one!" in every case, right?

For me it's okay to have just this one sentence in any case.

Further clarification is not necessarily necessary.
For example, if the Resurrection spell can't be used, it's because there are no troops to resurrect.
If the Blind spell can't be used, it's because the opposing troops are naturally immune, or the opposing hero has an artifact that protects his troops from the Blind spell.

If you see any specific cases that could lead to confusion and require further explanation, please specify them.
Please feel free to make suggestions for the text in these cases, as I haven't really thought about it at the moment, considering the standard sentence to be sufficient. :)

It's also important to remember that there's also translation work to be done for each new sentence added in English. :)
Now, if the translators are enthusiastic, it would also be possible to add the exact reason why the spell can't be used,
I'm not against that. :)

**Edit:**

I think that for specific cases, a more explanatory text could be added.
For example, for the Bless spell, there's no point in casting it on peasants because this creature doesn't have a damage range (e.g. "Damage: "1-3"), its damage is always fixed at "1".

**Comment by Kazhuu** [comment]
> There is no advantage to cast the "Bless" spell, or the "Curse" spell, or the "Slow" spell if the only troop is a troop of Peasants, so it should not be possible to cast either of these spells in this case.

When opposing side only has peasants. Slow and mass slow spells affect the movement of the unit. So I think it should be possible to cast it.

**Reply by LeHerosInconnu** [reply]
Hello @Kazhuu,

> > There is no advantage to cast the "Bless" spell, or the "Curse" spell, or the "Slow" spell if the only troop is a troop of Peasants, so it should not be possible to cast either of these spells in this case.
> 
> When opposing side only has peasants. Slow and mass slow spells affect the movement of the unit. So I think it should be possible to cast it.

Yes, my bad, you're right about the Slow spells. :)
I'll modify the original message.

**Reply by Alucard648** [reply]
Resurrection and True Ressurection, when there are no dead troops in non-elemental stacks: "There are no dead troops in your army, the spell is useless... for now."

Animate Dead when there are no destroyed undead creatures in stacks: "There are no fallen undead troops in your army, the spell is useless... for now."

Death Ripple and Death Wave spells, when all troops in enemy army are undead: "This spell will cause death and decay to your army and do nothing to your enemy!"

Holy Word and Holy Shout when there are no undead troops in enemy army: "This spell will devastate your undead army and do nothing to your enemy!"

Antimagic spell when enemy hero has 0 mana and there are no spellcasting creatures in his army: "Enemy hero can`t cast any spells and none of his creatures can cast spells on attack. This spell is useless for now."

Earthquake spell outside siege battles: "There are no castle walls to destroy using this spell"

Earthquake spell during castle defending: "This spell will devastate your castle walls!"

Summon Elementals spells when other summoned elementals are present on battlefield: "Those elemtal beings will kill each other, if not completely controlled."

Dispel and Mass Dispel, when there are no spells affecting your or enemy troops: "Currently there are no magic effects affecting anyone on this battlefield, this spell is useless for now."

Shield and Mass Shield spells when sieging enemy castle when only shooting enemies are arrow towers: "Castle towers use enchanted crossbow bolts that can penetrate any kind of defense, physical or magical, including effect of this spell"
